File: cookbook/php/Imports/createManualImportOn2Steps.php
<?php

//Url de base
$urlBase = 'http://v8.mailperformance.com/';

//X-Key
$xKey = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';



function getCreateImportJson($idBinding)
{
    // Remplissez les informations obligatoires
    $importName = "Nom de votre Import Manuel"; //Nom de l'import

    $segmentId = 1234; // Id du segment
    $phoneFieldsId = 1234; // Id du field
    $contactsId = array("1234ABCD"); // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
    $groupsContactsId = array(1234); // Id des groupes : Administration -> Groupes -> Idenfitifant

    $ImportData = array(
        "name" => $importName,// Remplissez les informations obligatoires
        "features" => array(
            array( // Parametre de normalisation
                "type" => "normalization",
                "fields" => array(
                    $phoneFieldsId => array("FRA") // Pour changer de pays de normalisation utiliser l'ISO ALPHA-3 Code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3)
                )
            ),
            array( // Ajout des cibles a un Segment
                "type" => "segmentation",
                "segmentId" => $segmentId,
                "emptyExisitingSegment" => true
            ),
            /*array( // Ajout des cibles a la liste rouge
                "type" => "redList",
                "destination" => array(
                    "sms" => true,
                    "email" => true
                )
            ),*/
            array( // Regles a appliquer en cas de doublon
                "type" => "duplicate",
                "rules" => array( "ignore" => true ) /// ou "rules" => array( "first" => true ) ou "rules" => array( "last" => true )
            ),
            array( // Parametres du rapport
                "type" => "report",
                "sendFinalReport" => false,
                "sendErrorReport" => true,
                "contactGuids" => $contactsId, // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
                "groupIds" => $groupsContactsId, // Id des groupes : Administration -> Groupes -> Idenfitifant
            ),
            array( // Parametres pour mettre a jour une cible dans la base de donnees
                "type" => "database",
                "updateExisting" => true,
                "crushData" => false
            )
        ),
        "binding" =>  $idBinding,
    );

    $createImportJson = json_encode($ImportData);

    echo "json :\n" . $createImportJson . "\n";

    return ($createImportJson);
}

function getExecutionImportJson()
{
    // Remplissez les informations obligatoires
    $fromFile = "C:\\votre\\chemin\\vers\\le\\fichier\\";
    $nameFile = "nomDeVotreFichier.extension";

    $sourceImportData = array(
            "data" => file_get_contents($fromFile . $nameFile),
            "name" => $nameFile
    );
    $sourceImportJson = json_encode($sourceImportData);


    echo "json :\n" . $sourceImportJson . "\n";

    return ($sourceImportData);
}


//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DES CREATION DES JSONS///////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DU PROGRAMME///////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


$idBinding = 1234;

$req = NULL;

//Nouvelle url
$url = $urlBase . "imports";
//On cree l'import de base
echo "CREATION IMPORT\n";
$con = urlPostOrPut($req, 'POST', getCreateImportJson($idBinding), $xKey, $url, 0);
if (($req = verification($con)) == null)
    return (0);

$importId = json_decode($con['result'], true)['id'];

//Nouvelle url
$url = $urlBase . "imports/" . $importId . "/executions";
//On execute l'import
echo "EXECUTION IMPORT\n";
$con = urlPostOrPut($req, 'POST', getExecutionImportJson(), $xKey, $url, 1);
if (($req = verification($con)) == null)
    return (0);


curl_close($req);
return (1);


//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DU PROGRAMME/////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DES FONCTIONS//////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


function verification($con)
{
    $result = $con['result'];
    $info = $con['info'];
    $req = $con['req'];

    //Verification des reponses
    if ($result == false)
    {
        //Erreur lors de la requete

        //Affichage de l'erreur
        $info = curl_getinfo($req);
        echo 'Error : ' . $info['http_code'] . "\n" . $result;
        return (null);
    }
    else
    {
        //On affiche la reponse
        echo 'OK : ' . $result . "\n\n";
        return ($req);
    }
}



//Utilisation de cURL pour remplir les requetes
//Fonctions de connexion
function startCurlInit($url)
{
    echo 'URL = ' . $url . "\n";
    $init = curl_init();
    curl_setopt($init, CURLOPT_URL, $url);
    curl_setopt($init, CURLOPT_RETURNTRANSFER, true);
    return ($init);
}

function urlPostOrPut($req, $request, $dataJson, $xKey, $url, $source)
{
    //On remplit la requete avec le bon verbe ($request) : POST / PUT
    $req = startCurlInit($url);
    curl_setopt($req, CURLOPT_CUSTOMREQUEST, $request);

    //Mise en place du xKey et des options
    if ($source == 0)
    {
        //Envoie normal
        curl_setopt($req, CURLOPT_POSTFIELDS, $dataJson);
        curl_setopt($req, CURLOPT_HTTPHEADER, array(
            'X-Key: ' . $xKey,
            'Content-Type: application/json',
            'Accept: application/vnd.mperf.v8.import.v1+json',
            'Content-Length: ' . strlen($dataJson)
        ));
    }
    else if ($source == 1)
    {
        //Envoie des informations du fichier selectionne
        curl_setopt($req, CURLOPT_POSTFIELDS, $dataJson["data"]);
        curl_setopt($req, CURLOPT_HTTPHEADER, array(
                'X-Key: ' . $xKey,
                'Accept: application/vnd.mperf.v8.import.v1+json',
                'Content-Type: application/octet-stream',
                'Content-Disposition: form-data; filename=' . $dataJson["name"]
        ));

    }

    //Execution de la requete
    $result = curl_exec($req);

    //Verification des reponses
    $info = curl_getinfo($req);

    return (array('result' => $result, 'info' => $info, 'req' => $req));
}

?>

File: cookbook/java/Imports/createManualImportOn2Steps.java
import java.io.BufferedReader;
import java.io.DataOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class createManualImportOn2Steps
{
    public static JSONObject getCreateImportJson(int idBinding) throws JSONException
    {
        // Remplissez les informations obligatoires
        String importName = "Nom de votre Import Manuel"; //Nom de l'import

        int segmentId = 1234; // Id du segment
        String phoneFieldsId = "1234"; // Id du field
        String contactsId[] = {"1234ABCD"}; // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
        int groupsContactsId[] = { 1234 }; // Id des groupes : Administration -> Groupes -> Idenfitifant


        String phoneFieldsIdArray[] = {"FRA"}; // Pour changer de pays de normalisation utiliser l'ISO ALPHA-3 Code (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-3)

        JSONObject fields = new JSONObject();
        fields.put(phoneFieldsId, phoneFieldsIdArray);

        JSONObject normalization = new JSONObject(); // Parametre de normalisation
        normalization.put("type", "normalization");
        normalization.put("fields", fields);

        JSONObject segmentation = new JSONObject(); // Ajout des cibles a un Segment
        segmentation.put("type", "segmentation");
        segmentation.put("segmentId", segmentId);
        segmentation.put("emptyExisitingSegment", true);

        JSONObject destination = new JSONObject();
        destination.put("sms", true);
        destination.put("email", true);

        JSONObject redList = new JSONObject(); // Ajout des cibles a la liste rouge
        redList.put("type", "redList");
        redList.put("destination", destination);

        JSONObject rules = new JSONObject();
        rules.put("ignore", true); // ou rules.put("first", true); ou rules.put("last", true);

        JSONObject duplicate = new JSONObject(); // Regles a appliquer en cas de doublon
        duplicate.put("type", "duplicate");
        duplicate.put("rules", rules);

        JSONObject report = new JSONObject(); // Parametres du rapport
        report.put("type", "report");
        report.put("sendFinalReport", false);
        report.put("sendErrorReport", true);
        report.put("contactGuids", contactsId); // Id des utilisateurs : Administration -> Utilisateurs -> Idenfitifant
        report.put("groupIds", groupsContactsId); // Id des groupes : Administration -> Groupes -> Idenfitifant

        JSONObject database = new JSONObject(); // Parametres pour mettre a jour une cible dans la base de donnees
        database.put("type", "database");
        database.put("updateExisting", true);
        database.put("crushData", false);

        JSONArray features = new JSONArray();
        features.put(normalization);
        features.put(segmentation);
        //features.put(redList);
        features.put(duplicate);
        features.put(report);
        features.put(database);

        JSONObject createImportJson = new JSONObject();
        createImportJson.put("name", importName);
        createImportJson.put("features", features);
        createImportJson.put("binding", idBinding);

        //On affiche le Json
        System.out.print("Json :\n" + createImportJson + "\n");

        return (createImportJson);
    }

    public static String[] getFileSourceInfo() throws JSONException, IOException
    {
        // Remplissez les informations obligatoires
        String path = "C:\\votre\\chemin\\vers\\le\\fichier\\";
        String name = "nomDeVotreFichier.extension";
        String file = null;

        //On lit l'interieur du fichier
        BufferedReader br = new BufferedReader(new FileReader(path + name));
        try {
            StringBuilder sb = new StringBuilder();
            String line = br.readLine();

            while (line != null) {
                sb.append(line);
                sb.append(System.lineSeparator());
                line = br.readLine();
            }
            file = sb.toString();
        } finally {
            br.close();
        }

        String result[] = {file, name};

        return (result);
    }

//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DES CREATION DES JSONS///////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DU PROGRAMME///////////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


    public static void main(String[] args) throws Exception
    {
        HttpURLConnection con = null;
        String verif = null;


        //Url de base
        String urlBase = "http://v8.mailperformance.com/";
        //X-Key
        String xKey = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
        int idBinding = 1234;  //Bindings cree par les APIs : importFormats.

        //Lancement de la connexion pour remplir la requete
        String url = urlBase + "imports/";
        //On cree l'import de base
        System.out.print("CREATE IMPORT\n");
        con = urlPostOrPut(con, "POST", getCreateImportJson(idBinding), xKey, url);
        if ((verif = verification(con)) == null)
            System.exit(0);
        System.out.printf("OK : " + verif + "\n\n");
        int importId = (new JSONObject(verif)).getInt("id");


        //Nouvelle url
        url = urlBase + "imports/" + importId + "/executions";
        //On execute l'import
        System.out.print("EXECUTION IMPORT\n");
        con = sendFile(con, "POST", getFileSourceInfo(), xKey, url);
        if ((verif = verification(con)) == null)
            System.exit(0);
        System.out.printf("OK : " + verif + "\n\n");



        con.disconnect();
    }


/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////FIN DU PROGRAMME////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////DEBUT DES FONCTIONS/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////


    public static String verification(HttpURLConnection con) throws IOException, JSONException
    {
        //Verification des reponses
        int responseCode = con.getResponseCode();
        if (responseCode != 200 && responseCode != 204)
        {
            //Affichage de l'erreur
            System.out.print("Error : " + responseCode + ' ' + con.getResponseMessage() + ".");
            return (null);
        }
        return (readConnection(con));
    }

    public static String readConnection(HttpURLConnection con) throws IOException, JSONException
    {
        //Lecture des donnees
        BufferedReader buffRead = new BufferedReader(new InputStreamReader(con.getInputStream()));
        String reply = buffRead.readLine();
        buffRead.close();

        return (reply);
    }



    //Fonctions de connexion
    public static HttpURLConnection startCon(String url, String xKey) throws MalformedURLException, IOException
    {
        //Lancement de la connexion
        HttpURLConnection con = (HttpURLConnection)new URL(url).openConnection();

        //Mise en place de la xKey et des options
        con.setRequestProperty("X-Key", xKey);
        con.setRequestProperty("Content-Type", "application/json");
        con.setRequestProperty("Accept", "application/vnd.mperf.v8.import.v1+json");
        return (con);
    }

    public static HttpURLConnection urlPostOrPut(HttpURLConnection con, String request, JSONObject dataJson, String xKey, String url) throws MalformedURLException, IOException, JSONException
    {
        //Lancement de la connexion pour remplir la requete
        con = startCon(url, xKey);
        con.setRequestProperty("Content-Length", Integer.toString(dataJson.toString().length()));
        con.setRequestMethod(request);
        con.setDoOutput(true);

        // Envoie des informations dans la connexion
        DataOutputStream streamCon = new DataOutputStream(con.getOutputStream());
        streamCon.write(dataJson.toString().getBytes());
        streamCon.flush();
        streamCon.close();

        return (con);
    }

    public static HttpURLConnection sendFile(HttpURLConnection con, String request, String fileInfo[], String xKey, String url) throws MalformedURLException, IOException
    {
        String file = fileInfo[0];
        String nameFile = fileInfo[1];

        System.out.printf("\nnamefile = " + nameFile + "\nFILE = " + file + "\n");

        con = (HttpURLConnection)new URL(url).openConnection();
        con.setDoOutput(true);
        con.setRequestMethod(request);
        con.setRequestProperty("X-Key", xKey);
        con.setRequestProperty("Accept", "application/vnd.mperf.v8.import.v1+json");
        con.setRequestProperty("Content-Type", "application/octet-stream");
        con.setRequestProperty("Content-Disposition", "form-data; filename=" + nameFile);

        // Envoie des informations dans la connexion
        DataOutputStream streamCon = new DataOutputStream(con.getOutputStream());
        streamCon.write(file.getBytes());
        streamCon.flush();
        streamCon.close();

        return (con);
    }
}
